contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,876 @@
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address consensusAddress) external payable onlyInitialized onlyValidatorContract {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        Validator memory valInfo = _validators[operatorAddress];
+        if (valInfo.poolModule == address(0) || valInfo.jailed) {
+            emit RewardDistributeFailed(operatorAddress, "INVALID_VALIDATOR");
+            return;
+        }
+
+        try IStakePool(valInfo.poolModule).distributeReward{ value: msg.value }(valInfo.commission.rate) {
+            emit RewardDistributed(operatorAddress, msg.value);
+        } catch Error(string memory reason) {
+            emit RewardDistributeFailed(operatorAddress, bytes(reason));
+        } catch (bytes memory lowLevelData) {
+            emit RewardDistributeFailed(operatorAddress, lowLevelData);
+        }
+    }
+
+    /**
+     * @dev Get new eligible validators from consensus engine
+     */
+    function updateEligibleValidators(
+        address[] calldata validators,
+        uint64[] calldata votingPowers
+    ) external onlyInitialized onlyCoinbase onlyZeroGasPrice {
+        uint256 newLength = validators.length;
+        if (newLength == 0) {
+            return;
+        }
+        uint256 oldLength = _eligibleValidators.length;
+        if (oldLength > newLength) {
+            for (uint256 i = newLength; i < oldLength; ++i) {
+                _eligibleValidators.pop();
+                _eligibleValidatorVoteAddrs.pop();
+            }
+        }
+
+        uint256 j;
+        for (uint256 i; i < newLength; ++i) {
+            address consensusAddress = validators[i];
+            address operatorAddress = _consensusToOperator[consensusAddress];
+            if (operatorAddress == address(0)) {
+                continue;
+            }
+
+            if (j >= oldLength) {
+                _eligibleValidators.push(
+                    IBSCValidatorSet.Validator({
+                        consensusAddress: consensusAddress,
+                        feeAddress: payable(0),
+                        BBCFeeAddress: address(0),
+                        votingPower: votingPowers[i],
+                        jailed: false,
+                        incoming: 0
+                    })
+                );
+                _eligibleValidatorVoteAddrs.push(_validators[operatorAddress].voteAddress);
+            } else {
+                IBSCValidatorSet.Validator storage eVal = _eligibleValidators[j];
+                eVal.consensusAddress = consensusAddress;
+                eVal.votingPower = votingPowers[i];
+                _eligibleValidatorVoteAddrs[j] = _validators[operatorAddress].voteAddress;
+            }
+            ++j;
+        }
+    }
+
+    function downtimeSlash(address consensusAddress, uint256 height) external onlyInitialized onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
```

Comment:
It is possible that a validator can edit the consensus addr to avoid downtime slash